Update LDT/DeVeny + "quality of life" changes

tbowers7 opened the pull request:
Primarily an update to LDT/DeVeny:
* Default parameter changes (mainly edge tracing and wavelength solution orders)
* Instrument-Specific arc line lists
* Updated wavelength template for one grating

A series of small "quality of life" improvements:
* Allow `pypeit_chk_wavecalib` to accept lists of input files.
* Allow restricted wavelength range for spectral flexure cross-correlation (this required touching a number of files in order to plumb the user-specifiable parameter all the way down to the function where it is used). This feature is useful when the end(s) of the wavelength calibration are not well constrained because of arc line placement and the user doesn't want those regions to affect the flexure correction.
* Update verbosity assignment in `pypeit_show_2dspec`.
* Doc/param fix to align with code for polynomial overscan fitting; update `np.polyfit()` -> `np.polynomial.Chebyshev.fit()`.

Main repo unit tests pass, will request cloud run of DevSuite following review.
```
================= test session starts ================
platform darwin -- Python 3.11.6, pytest-7.4.2, pluggy-1.3.0
rootdir: /Users/tbowers/d1/codes/PypeIt
configfile: setup.cfg
testpaths: pypeit/tests
plugins: doctestplus-1.0.0, astropy-0.11.0, hypothesis-6.88.1, remotedata-0.4.1, asdf-3.0.0, cov-4.1.0, filter-subpackage-0.1.2, astropy-header-0.2.2, mock-3.12.0, arraydiff-0.5.0
collected 240 items
...
==== 240 passed, 67 warnings in 408.87s (0:06:48) ====
```

Companion DevSuite PR (https://github.com/pypeit/PypeIt-development-suite/pull/295) adds wavelength templates.

tbowers7 committed:
DeVeny slit edge adjustments

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Update DeVeny Hg linelist; adjust rtol and continuum_subtract

	modified:   pypeit/data/arc_lines/lists/Hg_DeVeny1200_lines.dat
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Collection of sundry changes:

* Allow `pypeit_chk_wavecalib` to accept lists of input files

If more than one WaveCalib is to be checked at once, this commit allows for
the script to accept a list of input files, and simply loops the main execution
over the list of input files.

* Allow restricted wavelength range for flexure cross-correlation

Add FlexurePar parameters ``minwave`` and ``maxwave`` that may be used to
restrict the wavelength range used for the flexure cross-correlation.  This is
useful in the cases where the wavelength calibration _at the edges_ of the
spectral range is poorly constrained, yet there is a night-sky line used in
the correlation there.

Other uses may be possible.

The new parameters are passed to the wavelength restrictor in
`pypeit.core.flexure.spec_flux_shift` as 0 and numpy.inf as the effective min
and max specified values, respectively, (while the actual parameters are
``None`` if not specified).

* Tweak Hg wavelength

* Adjust wavecal fit parameters for LDT/DeVeny

* Update verbosity assignment in show_2dspec

	modified:   pypeit/core/flexure.py
	modified:   pypeit/data/arc_lines/lists/Hg_DeVeny1200_lines.dat
	modified:   pypeit/extraction.py
	modified:   pypeit/find_objects.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/scripts/chk_wavecalib.py
	modified:   pypeit/scripts/show_2dspec.py
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Align docs/params to code reality in overscan poly

The code for fitting a polynomial to the overscan region of a raw
image file for subtraction only requires one parameter (order), whereas
the documentation and PypeIt parameter fed into the overscan-subtraction
routine say "three".  This commit aligns the documentation and params
with the reality of the code.  Also, for the polynomial fitting, move
away from the `np.polyfit()` function (deprecated in NumPy v1.4 -- way
way long time ago) to the `numpy.polynomial` pagkage.  In particular,
use a Chebyshev polynomial for the fitting here (orthonormal basis sets
can be useful for seeing where the power is).

Modify the LDT/DeVeny params to use a polynomial (order=1) for fitting
the overscan.  This is more appropriate for this chip and EMI pickup
environment than the default SavGol filter.

	modified:   pypeit/core/procimg.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Merge branch 'develop' into update_deveny

tbowers7 committed:
Release Notes

	modified:   doc/releases/1.14.1dev.rst

tbowers7 committed:
Create / update DeVeny-specific line lists

Neon lines routinely recovered with LDT/DeVeny

These lines were originally included with #1595, but were causing various issues
with DevSuite tests.  Once more robust DevSuite VET tests are developed for
wavelength calibration checking, we can try again with these neon lines.

	new file:   pypeit/data/arc_lines/lists/ArI_DeVeny_lines.dat
	renamed:    pypeit/data/arc_lines/lists/Cd_DeVeny1200_lines.dat -> pypeit/data/arc_lines/lists/CdI_DeVeny_lines.dat
	renamed:    pypeit/data/arc_lines/lists/Hg_DeVeny1200_lines.dat -> pypeit/data/arc_lines/lists/HgI_DeVeny_lines.dat
	new file:   pypeit/data/arc_lines/lists/NeI_DeVeny_lines.dat
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Update the DV5 reid_arxiv template for LDT/DeVeny

Add function in the template-maker file for DV5.

(Update things to pathlib.Path... needed to add a force-str in one of
the other functions.)

	modified:   pypeit/core/wavecal/spectrographs/templ_ldt_deveny.py
	modified:   pypeit/core/wavecal/templates.py
	modified:   pypeit/data/arc_lines/reid_arxiv/ldt_deveny_500_HgCdAr.fits

kbwestfall commented on pypeit/core/flexure.py:
```diff
@@ -141,6 +141,14 @@ def spec_flex_shift(obj_skyspec, arx_skyspec, arx_fwhm_pix, spec_fwhm_pix=None,
             Two methods are available: 'boxcar' and 'slitcen' (see spec_flexure_slit()).
             In this routine, 'method' is only passed to final dict.
+        minwave (:obj:`float`, optional):
+            Miniuim wavelength to use for the correlation.  If ``None`` or less than
```
Minimum

kbwestfall commented on pypeit/core/flexure.py:
```diff
@@ -141,6 +141,14 @@ def spec_flex_shift(obj_skyspec, arx_skyspec, arx_fwhm_pix, spec_fwhm_pix=None,
+            the minumum wavelength of either ``obj_skyspec`` or ``arx_skyspec``,
+            this has no effect.  Default is None.
+        maxwave (:obj:`float`, optional):
+            Maximum wavelength to use for the correlation.  If ``None`` or grater than
```
greater

kbwestfall commented on pypeit/core/flexure.py:
```diff
@@ -432,6 +442,14 @@ def spec_flex_shift_global(slit_specs, islit, sky_spectrum, arx_fwhm_pix, empty_
+        maxwave (:obj:`float`, optional):
+            Maximum wavelength to use for the correlation.  If ``None`` or grater than
+            the maximum wavelength of either this sky or ``sky_spectrum``,
+            this has no effect.  Default is None.
```
Same typos from above occur here and below.

kbwestfall commented on pypeit/core/procimg.py:
```diff
@@ -672,9 +672,8 @@ def subtract_overscan(rawframe, datasec_img, oscansec_img, method='savgol', para
-            c = np.polyfit(np.arange(osfit.size), osfit, params[0])
-            ossub = np.polyval(c, np.arange(osfit.size))
+            poly = np.polynomial.Chebyshev.fit(np.arange(osfit.size), osfit, params[0])
+            ossub = poly(np.arange(osfit.size))
```
I suggest that we make this more explicit.  For now I would do:
```
if method.lower() == 'polynomial':
    warnings.warn('Method "polynomial" is identical to "chebyshev".  Former will be deprecated.', DeprecationWarning)
if method.lower() in ['polynomial', 'chebyshev']:
    ...
```

And then eventually we get rid of the polynomial option.

tbowers7 replied:
I added the suggested code.  When I run this with `method=polynomial`, however, I am not seeing the `DeprecationWarning` (_i.e._, it is being filtered somewhere).  I couldn't find a warning filter within PypeIt, so it may be coming from one of the imported libraries.  In this case, do we change this to a `msgs.warn()`, or just let it be for now?  I've included the same notice in the `pypeitpar` listing, so it will show up in the docs.

kbwestfall replied:
D'Oh!  Ah, well.  Those warnings would likely fly by unnoticed anyway.  I think having the notice in the docs is likely more useful anyway!

kbwestfall commented on pypeit/par/pypeitpar.py:
```diff
@@ -268,10 +268,9 @@ def __init__(self, trim=None, apply_gain=None, orient=None,
-                                'number of repeats ; for \'savgol\', set overscan_par = ' \
-                                'order, window size ; for \'median\', set overscan_par = ' \
-                                'None or omit the keyword.'
+                                '\'polynomial\', set overcan_par = order; for \'savgol\', ' \
```
change "polynomial" to "chebyshev"

tbowers7 replied:
Done.  Added deprecation warning here, and also included both "chebyshev" and "polynomial" in this class's `valid_overscan_methods()` method to allow both for now.

kbwestfall commented on pypeit/scripts/chk_wavecalib.py:
```diff
@@ -15,44 +15,45 @@ def get_parser(cls, width=None):
                                     width=width)
 
-        parser.add_argument('input_file', type=str,
+        parser.add_argument('input_file', type=str, nargs='+',
```
Edit the help string to indicate more than one file can be provided.

tbowers7 replied:
Done.

kbwestfall commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -637,6 +631,43 @@ def get_rawimage(self, raw_file, det):
         # Return
         return detector, raw_img, hdu, exptime, rawdatasec_img, oscansec_img
 
+    def calc_pattern_freq(self, frame, rawdatasec_img, oscansec_img, hdu):
```
I think this points to a change that we should make in `rawimage.RawImage.subtract_pattern` instead.  I.e., I'm not sure why that function should continue to attempt to subtract the pattern if this function returns an empty list.  Maybe @rcooke-ast has some advice?

rcooke-ast replied:
This definition is currently the same as the parent class, with the exception of the todo item that is added. Perhaps the todo item can be added to the top of the child class, unless there is intent to calculate the pattern frequency?

EDIT : I really appreciate the link to the EMI pickup description! If you want to chat more about how to remove this signal, we can chat about it some time.

tbowers7 replied:
The TODO is mostly for myself, with this stub method being a placeholder.  I just completed work on an external EMI pattern scrubber, and I placed this here with an eye toward (eventually) seeing how that compares to `rawimage.RawImage.subtract_pattern`.

kbwestfall reviewed (approved):
Thanks, Tim!  I have a few very minor comments and typo fixes.

I think your change from a warning to an error for the pattern-noise subtraction likely points to a deeper change we should make to `rawimage`, but I'd like to get input from @rcooke-ast about it.  I will ping him over Slack.

rcooke-ast commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -268,17 +269,19 @@ def default_pypeit_par(cls):
         par['calibrations']['arcframe']['process']['clip'] = False
         par['calibrations']['arcframe']['process']['combine'] = 'mean'
-        par['calibrations']['arcframe']['process']['subtract_continuum'] = True
+        # par['calibrations']['arcframe']['process']['subtract_continuum'] = True
```
Is there a reason to uncomment this for the arc lines? To be fair, it's only needed in cases where arc frames are being that contain different lamp spectra. Is this the case? Looking at the dev suite, it seems that dv7 contains multiple different lamp spectra?

tbowers7 replied:
The continuum-subtraction is an aesthetic issue for me -- without the subtraction (and hence negative flux values), the WaveCalib QA plots are done with a log scale.  For most of our users, the desired lamps are all on at the same time (_e.g._, HgCdAr) -- but some insist on taking separate frames for each lamp.  I have instructions in our instrument manual that if arcs with individual arc lamps are used/combined, then this parameter needs to be set to `True`.

I should check the DV7 `.pypeit` file in the DevSuite -- thanks for flagging that.

tbowers7 replied:
Note: I fixed the DV7 `.pypeit` file in pypeit/Pypeit-development-suite#296.

rcooke-ast replied:
Just FYI - I ran the dev suite, and got an error, but I suspect that you have fixed this in this PR
<img width="562" alt="image" src="https://github.com/pypeit/PypeIt/assets/12585758/bc9148ed-7a58-49ed-b255-e73d90ba2c5a">

rcooke-ast reviewed (commented):
Some comments/thoughts  :-)

tbowers7 committed:
PR Comments

Fixed typos, added "chebyshev" option for overscan subtraction with a
deprecation notice for the "polynomial" option, and updated the argument
help for `pypeit_chk_wavecalib`.

	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/procimg.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/scripts/chk_wavecalib.py

tbowers7 committed:
Merge branch 'develop' into update_deveny

tbowers7 committed:
Update release notes to reflect PR comments

	modified:   doc/releases/1.14.1dev.rst

tbowers7 committed:
Another sweep for typos

	modified:   pypeit/core/procimg.py
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
Clarify `pypeit_chk_noise_2dspec` option help

Cleaned up the help string for the `--mode` command-line option.

	modified:   pypeit/scripts/chk_noise_2dspec.py

tbowers7 committed:
Doc update

Update docs realted to both this PR and the datacube PR.

NOTE: python files edited are docstring-only changes to make
sphinx happy, with the following exceptions:
  pypeit/core/procimg.py  & pypeit/spectrographs/ldt_deveny.py
  -- finish adding "chebyshev"

In setup.cfg, added `psutil` as a [test] and [dev] dependency for
use with the DevSuite (dependency added in
pypeit/Pypeit-development-suite#235).

	new file:   doc/api/pypeit.coadd3d.rst
	modified:   doc/api/pypeit.rst
	modified:   doc/help/pypeit_chk_noise_2dspec.rst
	modified:   doc/help/pypeit_chk_wavecalib.rst
	modified:   doc/include/class_datamodel_datacube.rst
	modified:   doc/include/dependencies_table.rst
	modified:   doc/pypeit_par.rst
	modified:   doc/scripts/build_datacontainer_datamodels.py
	modified:   pypeit/coadd3d.py
	modified:   pypeit/core/datacube.py
	modified:   pypeit/core/procimg.py
	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/specutils/pypeit_loaders.py
	modified:   setup.cfg

tbowers7 commented:
@kbwestfall:
Updated docs related to both this PR and the datacube PR.

**NOTE:** python files edited in [54afea0](https://github.com/pypeit/PypeIt/pull/1712/commits/54afea0573b3f17d85a6afce5a31b2a7a81e7174) are docstring-only changes to make sphinx happy, with the following exceptions:
  `pypeit/core/procimg.py`  & `pypeit/spectrographs/ldt_deveny.py` -- finish adding the "chebyshev" overscan method.

In `setup.cfg`, added the package `psutil` as a [test] and [dev] dependency for use with the DevSuite (dependency added in https://github.com/pypeit/PypeIt-development-suite/pull/235 but never updated here).

tbowers7 committed:
Update release notes

kbwestfall commented:
> @kbwestfall: Updated docs related to both this PR and the datacube PR.

Apologies.  Most (all?) of these doc fixes are repeats of things waiting on merging of other PRs (e.g., #1706) ...

> **NOTE:** python files edited in [54afea0](https://github.com/pypeit/PypeIt/pull/1712/commits/54afea0573b3f17d85a6afce5a31b2a7a81e7174) are docstring-only changes to make sphinx happy, with the following exceptions: `pypeit/core/procimg.py` & `pypeit/spectrographs/ldt_deveny.py` -- finish adding the "chebyshev" overscan method.

I'm being overly picky, but I noticed you re-ordered the imports in procimg.py.  I recommend we try to follow [PEP 8](https://peps.python.org/pep-0008/#imports).

> 
> In `setup.cfg`, added the package `psutil` as a [test] and [dev] dependency for use with the DevSuite (dependency added in [pypeit/PypeIt-development-suite#235](https://github.com/pypeit/PypeIt-development-suite/pull/235) but never updated here).

Thanks! (But also in #1706 )

tbowers7 commented:
> I'm being overly picky, but I noticed you re-ordered the imports in procimg.py. I recommend we try to follow [PEP 8](https://peps.python.org/pep-0008/#imports).

I think the new order is in PEP8 format:

- Standard library imports.  (`warnings`)
- Related third party imports.  (`astropy`, `numpy`, `scipy`)
- Local application/library specific imports.  (`pypeit`)

Am I misinterpreting something?

Sorry about all the duplicated changes -- was just trying to follow the [development workflow](https://pypeit.readthedocs.io/en/release/dev/development.html#workflow) and make things work.

kbwestfall commented:
> > I'm being overly picky, but I noticed you re-ordered the imports in procimg.py. I recommend we try to follow [PEP 8](https://peps.python.org/pep-0008/#imports).
> 
> I think the new order is in PEP8 format:
> 
>     * Standard library imports.  (`warnings`)
> 
>     * Related third party imports.  (`astropy`, `numpy`, `scipy`)
> 
>     * Local application/library specific imports.  (`pypeit`)
> 
> 
> Am I misinterpreting something?

No, sorry, I regretted that complaint almost immediately after sending it!  I mistakenly thought of numpy as a standard lib.  So, technically you're right.  E.g., astropy and numpy are both 3rd party.  But I think of these imports as a dependency tree with lowest dependencies first.  This means I tend to group numpy, scipy, + matplotlib first, then astropy, then ..., then pypeit.  I.e., astropy depends on numpy, but not vice versa.

> Sorry about all the duplicated changes -- was just trying to follow the [development workflow](https://pypeit.readthedocs.io/en/release/dev/development.html#workflow) and make things work.

Not your fault!  I was just bemoaning duplicated work.

tbowers7 committed:
Merge branch 'develop' into update_deveny

tbowers7 committed:
Merge branch 'develop' into update_deveny

rcooke-ast reviewed (approved):
Happy to approve, but it might be good to confirm that this passes the dev suite DV7 before merging. To be fair, the dev-suite fail is likely because that merge has been done into `develop` already, but the PypeIt one hasn't been merged.

tbowers7 commented:
> Happy to approve, but it might be good to confirm that this passes the dev suite DV7 before merging. To be fair, the dev-suite fail is likely because that merge has been done into `develop` already, but the PypeIt one hasn't been merged.

Yes, the DevSuite PR was merged into `develop` already, and the new DV7 `.pypeit` file there needs new line lists that are included in this PR.

I've requested a full cloud DevSuite run for this branch.

tbowers7 commented:
Apart from waiting for the DevSuite results, this PR needs to wait until #1706 clears because of duplicated doc changes that need merging.

tbowers7 committed:
Cosmetic changes from trying to understand wavecal RMS

No code changes other than variable name changes to make it easier to
follow what is being done in computing the RMS of the wavelength
calibration solutions.

Also, add instrument-specific documentation page.

	new file:   doc/spectrographs/ldt_deveny.rst
	modified:   doc/spectrographs/spectrographs.rst
	modified:   pypeit/core/wavecal/autoid.py
	modified:   pypeit/core/wavecal/wv_fitting.py
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Merge branch 'develop' into update_deveny

rcooke-ast commented:
Dev-suite passes 👍 
<img width="615" alt="image" src="https://github.com/pypeit/PypeIt/assets/12585758/b7df1008-6317-4dcb-afc5-65993340256f">

tbowers7 committed:
FWHM tewak for ldt_deveny

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Merge branch 'platescale_hotfix' into update_deveny

tbowers7 committed:
More small parameter/comment tweaks

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Merge branch 'develop' into update_deveny

tbowers7 committed:
Late comment from #1724

	modified:   README.rst

tbowers7 committed:
Merge branch 'develop' into update_deveny

tbowers7 committed:
Remake docs after merge

	modified:   doc/pypeit_par.rst
	modified:   setup.cfg
